### PR TITLE
Allow the addition of placeholder files (updated)

### PR DESCRIPTION
**Pull request by mbraak**
Based on #278

### CODE REVIEW

**Comment by BoPeng** [comment]
> I'm now going to see if there is an alternative to the .placeholder suffix in the url.

It is easy to pass `is_placeholder` to the frontend via `get_values()`, then perhaps you can do something to make sure `uploads[idx]` is clearly not a proper URL?

**Comment by mbraak** [comment]
@BoPeng I  added `placeholder` to `get_values` and that works. The javascript uses that attribute instead of the suffix.

There is one thing that doesn't work yet. I tested this in the test project on the `/placeholder` url:

* open the form in the browser
* click 'delete' on the placeholder file
  -> file is correctly deleted
* leave 'title' field empty and submit form
  -> message 'title' is required
  -> Placeholder file is back again: this is not what I expected

**Comment by mbraak** [comment]
I also changed the parameters of `PlaceholderUploadedFile`. But I'm not sure about those yet.

File: frontend/file_form.js [hunk 1]
```diff
@@ -323,16 +329,28 @@ class UploadFile {
   };
 
   handleDelete(uploadIndex) {
+    const { placeholder } = this.uploads[uploadIndex];
+
+    if (placeholder) {
+      this.deleteUpload(uploadIndex);
+    }
```
Comment (BoPeng) [inline frontend/file_form.js]: Did you miss the following?

```
else {
   this.deleteFromServer(uploadIndex);
}
```

**Comment by BoPeng** [comment]
I suspect that we have to persist the placeholders to the database anyway. Let me experiment this a bit more.

**Comment by BoPeng** [comment]
https://github.com/django/django/blob/9358da704ea9ba55f22df912e47b54eb85d5c97e/django/forms/forms.py#L388-L390

https://github.com/django/django/blob/9358da704ea9ba55f22df912e47b54eb85d5c97e/django/forms/fields.py#L579-L596

I think the problem is with these two pieces of code, which, with our PR, basically means,

1. if there are new files, initial files are ignored. This is **not** what we want. (to test this, add a new file, submit, you will only see the new file). We could potentially override `clean` to return `value + initial`.

2. if there is no new file, `clean()` is called with `[], initial_files`, and the initial files are returned, regardless if it has been removed from interface. This is also wrong. (To test, remove the file from interface, check the `initial` from `clean(self, value, initial)`, the `initial` is still there.

So, assuming providing `value` in the following line of code is the only way to get the initial file list in the widget,

https://github.com/mbraak/django-file-form/blob/master/django_file_form/widgets.py#L40

we will have to

1. when `value` is provided, create fake database records. Otherwise we will not be able to return the correct file list since the removal of placeholders will not change either new file list or initial value. The problem is that we do not know `form_id` now, right?
2. the frontend is unchanged, the delete action is trigged for placeholder files.
3. redefine `def clean(value, initial)` so that when there is no `value`, `initial` is not used.

I will try to implement these tomorrow.

**Comment by mbraak** [comment]
Perhaps it's also possible to use a hidden input to fix this. I'm going to see if that's possible.

**Comment by BoPeng** [comment]
Thanks. I also think the `bound_data` function of `MultipleUploadedFileField` needs a second look because it combines `value` and `initial` and is causing the `initial` data to be always re-displayed in the `delete`, `submit`, `re-render` due to lack of `title` scenario.

https://github.com/mbraak/django-file-form/blob/cf4b2a6ca1bec9aab72f568be6788c4d71eb3389/django_file_form/fields.py#L60-L69

**Comment by BoPeng** [comment]
@mbraak I am out of time today but I think we will have to do the following

1. In `upload_widget.html`, add something like

```
    <input type="hidden" id="placeholder_files" name="placeholder"
        value="">
```

2. In `file_form.js`, add to `addInitialFiles` the following
```
if (file.placeholder) {
        this.uploads.push({ placeholder: true });
        var placeholder_files = document.getElementById('placeholder_files');
        ... add file info to placeholder_files ...
```
then add to `handleDelete` to remove info from this hidden field

```
    if (placeholder) {
      this.deleteUpload(uploadIndex);
      var placeholder_files = document.getElementById('placeholder_files');
      placeholder_files.value = ' ... ';
    }
```

3. In `UploadMultipleWidget` in `widgets.py` check the `data` that is returned from the frontend, and add `placeholder` files to the file list.

4. Maybe, in `fields.py, change 

```
    def bound_data(self, data, initial):
        return data
```
because I think this function is only called when the form is re-rendered and at that time `initial` should not be used. Perhaps `get_list(data) if data else get_list(initial)` can accommodate other scenarios.

5. The same to `clean` in `MultipleUploadedFileField`. The initial was returned if there is no value. I think this is ok for single file case when users have not couched the widget, but in our case we should ignore `initial`. Not quite sure about this either.

**Comment by mbraak** [comment]
I added pr #283 with a first attempt at using hidden inputs.

**Comment by mbraak** [comment]
I merged some prs. Let's now see if this pr is complete.

**Comment by BoPeng** [comment]
Thanks. I did more testing and everything seems to be working correctly. I noticed that you changed `hasattr(f, 'placeholder')` to `hasattr(f,  'is_placeholder')`. I prefer the former because `has_attr` has a meaning of testing `is_` so the latter reads redundant. If you like `is_placeholder`, it is better to  add `is_placeholder` to `UploadedFilesWithId` so that the test becomes

```
if f.is_placeholder:
   continue
```
which is certainly better than `hasattr`.

Anyway, I do not have a strong opinion either way so I am ok with all three possibilities. Please pick one, update `README`, and merge the PR. I think a release is justified with the introduction of this major feature.

**Comment by mbraak** [comment]
Thanks! Ik will look at it.

**Comment by mbraak** [comment]
This looks complete now. I will see when I can merge this.

**Comment by BoPeng** [comment]
Thanks. What I have in mind, but definitely for later, is a customized `UploadedFileWithId` model that can be set in `settings` as an advanced feature that provided by many other django apps. I intended to add a field `description` to allow users to add a short description to the uploaded file. The backend is easy but the frontend is troublesome because the frontend does not know how to work with user-provided fields.

**Comment by mbraak** [comment]
merged!

**Pull request merged**